{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import glob\n",
    "import scipy\n",
    "import sklearn\n",
    "import matplotlib\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from collections import defaultdict\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "\n",
    "%matplotlib inline\n",
    "sns.set_style('whitegrid')\n",
    "pd.set_option('display.max_rows', 100)\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "pd.set_option('display.max_columns', 100)\n",
    "\n",
    "sns.set_palette(\"Set2\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataloc = '/LAB_DATA/CURRENT/CURRENT_Metagenomics_PROJECTS/2021_IgASeq/tables/'\n",
    "zenodoloc = '/home/mattolm/user_data/Programs/MIGSeq_code/zenodo/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Tube_ID</th>\n",
       "      <th>sample</th>\n",
       "      <th>IgA_frac</th>\n",
       "      <th>pass_sorting_QC</th>\n",
       "      <th>pass_SeqDepth_QC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1022</td>\n",
       "      <td>OS_62</td>\n",
       "      <td>0.4660</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1022</td>\n",
       "      <td>OS_24</td>\n",
       "      <td>0.1353</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1026</td>\n",
       "      <td>OS_64</td>\n",
       "      <td>0.3230</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1026</td>\n",
       "      <td>OS_26</td>\n",
       "      <td>0.0670</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1036</td>\n",
       "      <td>OS_54</td>\n",
       "      <td>0.5630</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Tube_ID sample  IgA_frac  pass_sorting_QC  pass_SeqDepth_QC\n",
       "2     1022  OS_62    0.4660             True              True\n",
       "3     1022  OS_24    0.1353             True              True\n",
       "4     1026  OS_64    0.3230             True              True\n",
       "5     1026  OS_26    0.0670             True              True\n",
       "6     1036  OS_54    0.5630             True              True"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# IgA binding data (IgA_frac); from supplemental table S1\n",
    "Idb = pd.read_csv(dataloc + 'SupplementalTable_S1_v1.csv')\n",
    "Idb = Idb[['Tube_ID', 'sample', 'IgA_frac', 'pass_sorting_QC', 'pass_SeqDepth_QC']].drop_duplicates()\n",
    "Idb = Idb[(Idb['pass_SeqDepth_QC'] == True) & (Idb['pass_sorting_QC'] == True)]\n",
    "display(Idb.head())\n",
    "\n",
    "# Raw genome abundance data; from Zenodo\n",
    "Gdb = pd.read_csv(zenodoloc + 'IgASeq_genomeDetection_v4.csv')\n",
    "Gdb = Gdb[Gdb['sample'].isin(set(Idb['sample']))]\n",
    "Gdb.head()\n",
    "\n",
    "assert set(Idb['sample']) == set(Gdb['sample'])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prepare to calculate IgA+ probability score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome</th>\n",
       "      <th>Tube_ID</th>\n",
       "      <th>positive_breadth</th>\n",
       "      <th>positive_rel_ab</th>\n",
       "      <th>positive_coverage</th>\n",
       "      <th>positive_sample</th>\n",
       "      <th>unsorted_breadth</th>\n",
       "      <th>unsorted_rel_ab</th>\n",
       "      <th>unsorted_coverage</th>\n",
       "      <th>unsorted_sample</th>\n",
       "      <th>IgA_frac</th>\n",
       "      <th>pass_sorting_QC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GUT_GENOME000139.fna.fa</td>\n",
       "      <td>7028</td>\n",
       "      <td>0.552788</td>\n",
       "      <td>0.000369</td>\n",
       "      <td>1.121264</td>\n",
       "      <td>OS_61</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OS_23</td>\n",
       "      <td>0.8070</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GUT_GENOME000139.fna.fa</td>\n",
       "      <td>7028</td>\n",
       "      <td>0.552788</td>\n",
       "      <td>0.000369</td>\n",
       "      <td>1.121264</td>\n",
       "      <td>OS_61</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OS_23</td>\n",
       "      <td>0.0767</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GUT_GENOME000205.fna.fa</td>\n",
       "      <td>7028</td>\n",
       "      <td>0.267296</td>\n",
       "      <td>0.000627</td>\n",
       "      <td>1.724991</td>\n",
       "      <td>OS_61</td>\n",
       "      <td>0.119370</td>\n",
       "      <td>0.000303</td>\n",
       "      <td>1.145421</td>\n",
       "      <td>OS_23</td>\n",
       "      <td>0.8070</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GUT_GENOME000205.fna.fa</td>\n",
       "      <td>7028</td>\n",
       "      <td>0.267296</td>\n",
       "      <td>0.000627</td>\n",
       "      <td>1.724991</td>\n",
       "      <td>OS_61</td>\n",
       "      <td>0.119370</td>\n",
       "      <td>0.000303</td>\n",
       "      <td>1.145421</td>\n",
       "      <td>OS_23</td>\n",
       "      <td>0.0767</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GUT_GENOME000221.fna.fa</td>\n",
       "      <td>7028</td>\n",
       "      <td>0.020986</td>\n",
       "      <td>0.000445</td>\n",
       "      <td>0.920143</td>\n",
       "      <td>OS_61</td>\n",
       "      <td>0.068223</td>\n",
       "      <td>0.001864</td>\n",
       "      <td>5.255976</td>\n",
       "      <td>OS_23</td>\n",
       "      <td>0.8070</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20443</th>\n",
       "      <td>METABAT215_SUBJECTMAPPING_SCAFFOLDS_min1500_CO...</td>\n",
       "      <td>8038</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OS_43</td>\n",
       "      <td>0.919023</td>\n",
       "      <td>0.000955</td>\n",
       "      <td>4.243006</td>\n",
       "      <td>OS_5</td>\n",
       "      <td>0.0860</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20444</th>\n",
       "      <td>METABAT215_SUBJECTMAPPING_SCAFFOLDS_min1500_CO...</td>\n",
       "      <td>8038</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OS_43</td>\n",
       "      <td>0.497457</td>\n",
       "      <td>0.000286</td>\n",
       "      <td>0.865073</td>\n",
       "      <td>OS_5</td>\n",
       "      <td>0.6760</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20445</th>\n",
       "      <td>METABAT215_SUBJECTMAPPING_SCAFFOLDS_min1500_CO...</td>\n",
       "      <td>8038</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OS_43</td>\n",
       "      <td>0.497457</td>\n",
       "      <td>0.000286</td>\n",
       "      <td>0.865073</td>\n",
       "      <td>OS_5</td>\n",
       "      <td>0.0860</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20446</th>\n",
       "      <td>METABAT215_SUBJECTMAPPING_SCAFFOLDS_min1500_CO...</td>\n",
       "      <td>8038</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OS_43</td>\n",
       "      <td>0.802729</td>\n",
       "      <td>0.000426</td>\n",
       "      <td>2.240798</td>\n",
       "      <td>OS_5</td>\n",
       "      <td>0.6760</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20447</th>\n",
       "      <td>METABAT215_SUBJECTMAPPING_SCAFFOLDS_min1500_CO...</td>\n",
       "      <td>8038</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OS_43</td>\n",
       "      <td>0.802729</td>\n",
       "      <td>0.000426</td>\n",
       "      <td>2.240798</td>\n",
       "      <td>OS_5</td>\n",
       "      <td>0.0860</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20448 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  genome  Tube_ID  \\\n",
       "0                                GUT_GENOME000139.fna.fa     7028   \n",
       "1                                GUT_GENOME000139.fna.fa     7028   \n",
       "2                                GUT_GENOME000205.fna.fa     7028   \n",
       "3                                GUT_GENOME000205.fna.fa     7028   \n",
       "4                                GUT_GENOME000221.fna.fa     7028   \n",
       "...                                                  ...      ...   \n",
       "20443  METABAT215_SUBJECTMAPPING_SCAFFOLDS_min1500_CO...     8038   \n",
       "20444  METABAT215_SUBJECTMAPPING_SCAFFOLDS_min1500_CO...     8038   \n",
       "20445  METABAT215_SUBJECTMAPPING_SCAFFOLDS_min1500_CO...     8038   \n",
       "20446  METABAT215_SUBJECTMAPPING_SCAFFOLDS_min1500_CO...     8038   \n",
       "20447  METABAT215_SUBJECTMAPPING_SCAFFOLDS_min1500_CO...     8038   \n",
       "\n",
       "       positive_breadth  positive_rel_ab  positive_coverage positive_sample  \\\n",
       "0              0.552788         0.000369           1.121264           OS_61   \n",
       "1              0.552788         0.000369           1.121264           OS_61   \n",
       "2              0.267296         0.000627           1.724991           OS_61   \n",
       "3              0.267296         0.000627           1.724991           OS_61   \n",
       "4              0.020986         0.000445           0.920143           OS_61   \n",
       "...                 ...              ...                ...             ...   \n",
       "20443          0.000000         0.000000           0.000000           OS_43   \n",
       "20444          0.000000         0.000000           0.000000           OS_43   \n",
       "20445          0.000000         0.000000           0.000000           OS_43   \n",
       "20446          0.000000         0.000000           0.000000           OS_43   \n",
       "20447          0.000000         0.000000           0.000000           OS_43   \n",
       "\n",
       "       unsorted_breadth  unsorted_rel_ab  unsorted_coverage unsorted_sample  \\\n",
       "0              0.000000         0.000000           0.000000           OS_23   \n",
       "1              0.000000         0.000000           0.000000           OS_23   \n",
       "2              0.119370         0.000303           1.145421           OS_23   \n",
       "3              0.119370         0.000303           1.145421           OS_23   \n",
       "4              0.068223         0.001864           5.255976           OS_23   \n",
       "...                 ...              ...                ...             ...   \n",
       "20443          0.919023         0.000955           4.243006            OS_5   \n",
       "20444          0.497457         0.000286           0.865073            OS_5   \n",
       "20445          0.497457         0.000286           0.865073            OS_5   \n",
       "20446          0.802729         0.000426           2.240798            OS_5   \n",
       "20447          0.802729         0.000426           2.240798            OS_5   \n",
       "\n",
       "       IgA_frac  pass_sorting_QC  \n",
       "0        0.8070             True  \n",
       "1        0.0767             True  \n",
       "2        0.8070             True  \n",
       "3        0.0767             True  \n",
       "4        0.8070             True  \n",
       "...         ...              ...  \n",
       "20443    0.0860             True  \n",
       "20444    0.6760             True  \n",
       "20445    0.0860             True  \n",
       "20446    0.6760             True  \n",
       "20447    0.0860             True  \n",
       "\n",
       "[20448 rows x 12 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Make a base table\n",
    "name = 'positive'\n",
    "db1 = Gdb[Gdb['IgA'] == name][['genome', 'Tube_ID', 'breadth', 'rel_ab', 'coverage', 'sample']].rename(columns={'breadth':name + '_breadth', 'rel_ab':name + '_rel_ab', 'coverage':name + '_coverage', 'sample':name + '_sample'})\n",
    "\n",
    "name = 'unsorted'\n",
    "db2 = Gdb[Gdb['IgA'] == \"resuspended_unsorted\"][['genome', 'Tube_ID', 'breadth', 'rel_ab', 'coverage', 'sample']].rename(columns={'breadth':name + '_breadth', 'rel_ab':name + '_rel_ab', 'coverage':name + '_coverage', 'sample':name + '_sample'})\n",
    "\n",
    "IGdb = pd.merge(db1, db2, on=['genome', 'Tube_ID'], how='outer')\n",
    "IGdb['Tube_ID'].value_counts()\n",
    "\n",
    "assert len(IGdb[IGdb.duplicated(subset=['genome', 'Tube_ID'], keep=False)].sort_values(['genome', 'Tube_ID'])) == 0\n",
    "\n",
    "# Figure out which NaNs should be 0s:\n",
    "for name in ['positive', 'unsorted']:\n",
    "    # Tube ID to sample\n",
    "    t2i = IGdb.dropna(subset=[name + '_sample']).set_index('Tube_ID')[name + '_sample'].to_dict()\n",
    "    \n",
    "    # Backfill to remove the NaNs\n",
    "    IGdb[name + '_sample'] = IGdb['Tube_ID'].map(t2i)\n",
    "\n",
    "    # Backfill to remove the 0s\n",
    "    for c in ['breadth', 'coverage', 'rel_ab']:\n",
    "        IGdb[name + '_' + c] = [0 if ((s == s) & (c_val != c_val)) else c_val for c_val, s in zip(IGdb[name + '_' + c], IGdb[name + '_sample'])]\n",
    "        \n",
    "# Add IgA binding\n",
    "IGdb = pd.merge(IGdb, Idb[['Tube_ID', 'IgA_frac', 'pass_sorting_QC']], how='left')\n",
    "assert set(IGdb['unsorted_sample']).union(set(IGdb['positive_sample'])) == set(Gdb['sample'])\n",
    "IGdb\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calulate IgA binding metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_iga_prob(row):\n",
    "    if row['unsorted_rel_ab'] == 0:\n",
    "        if row['positive_rel_ab'] > 0:\n",
    "            return 1e6\n",
    "        else:\n",
    "            return np.nan\n",
    "    return (row['positive_rel_ab'] * row['IgA_frac'] / row['unsorted_rel_ab'])\n",
    "\n",
    "IGdb['IgA_prob'] = IGdb.apply(calc_iga_prob, axis=1)\n",
    "\n",
    "IGdb['semi_robustly_detected'] = [(p >= 0.5) | (u >= 0.5) for p, u in zip(IGdb['positive_breadth'], IGdb['unsorted_breadth'])]\n",
    "IGdb['robustly_detected'] = [(p >= 0.5) & (u >= 0.5) for p, u in zip(IGdb['positive_breadth'], IGdb['unsorted_breadth'])]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Correct for sequencing depth\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analyzing 30 samples\n"
     ]
    }
   ],
   "source": [
    "THRESH = 12\n",
    "gs = 4000000\n",
    "rel_thresh = gs/(THRESH*1e9)\n",
    "\n",
    "# Make a filtered dataset that only include genomes that pass threse thresholds\n",
    "FIGdb = IGdb.copy()\n",
    "FIGdb = FIGdb[(FIGdb['positive_rel_ab'] >= (rel_thresh)) | (FIGdb['unsorted_rel_ab'] >= rel_thresh)]\n",
    "FIGdb = FIGdb[(FIGdb['positive_breadth'] >= 0.5) | (FIGdb['unsorted_breadth'] >= 0.5)]\n",
    "\n",
    "print(f\"Analyzing {len(FIGdb['Tube_ID'].unique())} samples\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0003333333333333333"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rel_thresh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set binding classes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome</th>\n",
       "      <th>Tube_ID</th>\n",
       "      <th>positive_breadth</th>\n",
       "      <th>positive_rel_ab</th>\n",
       "      <th>positive_coverage</th>\n",
       "      <th>positive_sample</th>\n",
       "      <th>unsorted_breadth</th>\n",
       "      <th>unsorted_rel_ab</th>\n",
       "      <th>unsorted_coverage</th>\n",
       "      <th>unsorted_sample</th>\n",
       "      <th>IgA_frac</th>\n",
       "      <th>pass_sorting_QC</th>\n",
       "      <th>IgA_prob</th>\n",
       "      <th>semi_robustly_detected</th>\n",
       "      <th>robustly_detected</th>\n",
       "      <th>IgA_binding_class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GUT_GENOME000139.fna.fa</td>\n",
       "      <td>7028</td>\n",
       "      <td>0.552788</td>\n",
       "      <td>0.000369</td>\n",
       "      <td>1.121264</td>\n",
       "      <td>OS_61</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OS_23</td>\n",
       "      <td>0.8070</td>\n",
       "      <td>True</td>\n",
       "      <td>1000000.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>high</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GUT_GENOME000139.fna.fa</td>\n",
       "      <td>7028</td>\n",
       "      <td>0.552788</td>\n",
       "      <td>0.000369</td>\n",
       "      <td>1.121264</td>\n",
       "      <td>OS_61</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OS_23</td>\n",
       "      <td>0.0767</td>\n",
       "      <td>True</td>\n",
       "      <td>1000000.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>high</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>GUT_GENOME000767.fna.fa</td>\n",
       "      <td>7028</td>\n",
       "      <td>0.826847</td>\n",
       "      <td>0.000752</td>\n",
       "      <td>2.513575</td>\n",
       "      <td>OS_61</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OS_23</td>\n",
       "      <td>0.8070</td>\n",
       "      <td>True</td>\n",
       "      <td>1000000.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>high</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>GUT_GENOME000767.fna.fa</td>\n",
       "      <td>7028</td>\n",
       "      <td>0.826847</td>\n",
       "      <td>0.000752</td>\n",
       "      <td>2.513575</td>\n",
       "      <td>OS_61</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OS_23</td>\n",
       "      <td>0.0767</td>\n",
       "      <td>True</td>\n",
       "      <td>1000000.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>high</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>GUT_GENOME094675.fna.fa</td>\n",
       "      <td>7028</td>\n",
       "      <td>0.917889</td>\n",
       "      <td>0.001756</td>\n",
       "      <td>6.808159</td>\n",
       "      <td>OS_61</td>\n",
       "      <td>0.922662</td>\n",
       "      <td>0.002606</td>\n",
       "      <td>13.991979</td>\n",
       "      <td>OS_23</td>\n",
       "      <td>0.8070</td>\n",
       "      <td>True</td>\n",
       "      <td>0.543689</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20439</th>\n",
       "      <td>METABAT215_SUBJECTMAPPING_SCAFFOLDS_min1500_CO...</td>\n",
       "      <td>8038</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OS_43</td>\n",
       "      <td>0.918784</td>\n",
       "      <td>0.000405</td>\n",
       "      <td>3.630133</td>\n",
       "      <td>OS_5</td>\n",
       "      <td>0.0860</td>\n",
       "      <td>True</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>uncoated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20442</th>\n",
       "      <td>METABAT215_SUBJECTMAPPING_SCAFFOLDS_min1500_CO...</td>\n",
       "      <td>8038</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OS_43</td>\n",
       "      <td>0.919023</td>\n",
       "      <td>0.000955</td>\n",
       "      <td>4.243006</td>\n",
       "      <td>OS_5</td>\n",
       "      <td>0.6760</td>\n",
       "      <td>True</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>uncoated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20443</th>\n",
       "      <td>METABAT215_SUBJECTMAPPING_SCAFFOLDS_min1500_CO...</td>\n",
       "      <td>8038</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OS_43</td>\n",
       "      <td>0.919023</td>\n",
       "      <td>0.000955</td>\n",
       "      <td>4.243006</td>\n",
       "      <td>OS_5</td>\n",
       "      <td>0.0860</td>\n",
       "      <td>True</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>uncoated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20446</th>\n",
       "      <td>METABAT215_SUBJECTMAPPING_SCAFFOLDS_min1500_CO...</td>\n",
       "      <td>8038</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OS_43</td>\n",
       "      <td>0.802729</td>\n",
       "      <td>0.000426</td>\n",
       "      <td>2.240798</td>\n",
       "      <td>OS_5</td>\n",
       "      <td>0.6760</td>\n",
       "      <td>True</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>uncoated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20447</th>\n",
       "      <td>METABAT215_SUBJECTMAPPING_SCAFFOLDS_min1500_CO...</td>\n",
       "      <td>8038</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OS_43</td>\n",
       "      <td>0.802729</td>\n",
       "      <td>0.000426</td>\n",
       "      <td>2.240798</td>\n",
       "      <td>OS_5</td>\n",
       "      <td>0.0860</td>\n",
       "      <td>True</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>uncoated</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7038 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  genome  Tube_ID  \\\n",
       "0                                GUT_GENOME000139.fna.fa     7028   \n",
       "1                                GUT_GENOME000139.fna.fa     7028   \n",
       "14                               GUT_GENOME000767.fna.fa     7028   \n",
       "15                               GUT_GENOME000767.fna.fa     7028   \n",
       "58                               GUT_GENOME094675.fna.fa     7028   \n",
       "...                                                  ...      ...   \n",
       "20439  METABAT215_SUBJECTMAPPING_SCAFFOLDS_min1500_CO...     8038   \n",
       "20442  METABAT215_SUBJECTMAPPING_SCAFFOLDS_min1500_CO...     8038   \n",
       "20443  METABAT215_SUBJECTMAPPING_SCAFFOLDS_min1500_CO...     8038   \n",
       "20446  METABAT215_SUBJECTMAPPING_SCAFFOLDS_min1500_CO...     8038   \n",
       "20447  METABAT215_SUBJECTMAPPING_SCAFFOLDS_min1500_CO...     8038   \n",
       "\n",
       "       positive_breadth  positive_rel_ab  positive_coverage positive_sample  \\\n",
       "0              0.552788         0.000369           1.121264           OS_61   \n",
       "1              0.552788         0.000369           1.121264           OS_61   \n",
       "14             0.826847         0.000752           2.513575           OS_61   \n",
       "15             0.826847         0.000752           2.513575           OS_61   \n",
       "58             0.917889         0.001756           6.808159           OS_61   \n",
       "...                 ...              ...                ...             ...   \n",
       "20439          0.000000         0.000000           0.000000           OS_43   \n",
       "20442          0.000000         0.000000           0.000000           OS_43   \n",
       "20443          0.000000         0.000000           0.000000           OS_43   \n",
       "20446          0.000000         0.000000           0.000000           OS_43   \n",
       "20447          0.000000         0.000000           0.000000           OS_43   \n",
       "\n",
       "       unsorted_breadth  unsorted_rel_ab  unsorted_coverage unsorted_sample  \\\n",
       "0              0.000000         0.000000           0.000000           OS_23   \n",
       "1              0.000000         0.000000           0.000000           OS_23   \n",
       "14             0.000000         0.000000           0.000000           OS_23   \n",
       "15             0.000000         0.000000           0.000000           OS_23   \n",
       "58             0.922662         0.002606          13.991979           OS_23   \n",
       "...                 ...              ...                ...             ...   \n",
       "20439          0.918784         0.000405           3.630133            OS_5   \n",
       "20442          0.919023         0.000955           4.243006            OS_5   \n",
       "20443          0.919023         0.000955           4.243006            OS_5   \n",
       "20446          0.802729         0.000426           2.240798            OS_5   \n",
       "20447          0.802729         0.000426           2.240798            OS_5   \n",
       "\n",
       "       IgA_frac  pass_sorting_QC        IgA_prob  semi_robustly_detected  \\\n",
       "0        0.8070             True  1000000.000000                    True   \n",
       "1        0.0767             True  1000000.000000                    True   \n",
       "14       0.8070             True  1000000.000000                    True   \n",
       "15       0.0767             True  1000000.000000                    True   \n",
       "58       0.8070             True        0.543689                    True   \n",
       "...         ...              ...             ...                     ...   \n",
       "20439    0.0860             True        0.000000                    True   \n",
       "20442    0.6760             True        0.000000                    True   \n",
       "20443    0.0860             True        0.000000                    True   \n",
       "20446    0.6760             True        0.000000                    True   \n",
       "20447    0.0860             True        0.000000                    True   \n",
       "\n",
       "       robustly_detected IgA_binding_class  \n",
       "0                  False              high  \n",
       "1                  False              high  \n",
       "14                 False              high  \n",
       "15                 False              high  \n",
       "58                  True               low  \n",
       "...                  ...               ...  \n",
       "20439              False          uncoated  \n",
       "20442              False          uncoated  \n",
       "20443              False          uncoated  \n",
       "20446              False          uncoated  \n",
       "20447              False          uncoated  \n",
       "\n",
       "[7038 rows x 16 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Classify binding classes\n",
    "def calc_class(iga_norm):\n",
    "    if iga_norm < 0.005:\n",
    "        return 'uncoated'\n",
    "    elif iga_norm < 1:\n",
    "        return 'low'\n",
    "    else:\n",
    "        assert iga_norm >= 1, iga_norm\n",
    "        return 'high'\n",
    "    \n",
    "FIGdb['IgA_binding_class'] = [calc_class(x) for x in FIGdb['IgA_prob']]\n",
    "FIGdb"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "conda_base",
   "language": "python",
   "name": "conda_base"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
